File: dotnet-stack/backend/Services/CacheService.cs
using System.Collections.Concurrent;
using System.Text.Json;
using StackExchange.Redis;

namespace WildernessStays.Api.Services;

/// <summary>
/// Redis-backed cache with a transparent in-memory fallback, so the API
/// stays fully functional when Redis is unavailable (parity with the Node backend).
/// </summary>
public class CacheService
{
    private readonly ILogger<CacheService> _logger;
    private readonly ConnectionMultiplexer? _redis;
    private readonly ConcurrentDictionary<string, (string Value, DateTime ExpiresAt)> _memory = new();

    public CacheService(IConfiguration config, ILogger<CacheService> logger)
    {
        _logger = logger;
        try
        {
            var options = ConfigurationOptions.Parse(config["Redis"] ?? "localhost:6379");
            options.AbortOnConnectFail = false;
            options.ConnectTimeout = 1500;
            _redis = ConnectionMultiplexer.Connect(options);
            _logger.LogInformation("Redis configured — caching enabled (falls back to memory if unreachable)");
        }
        catch (Exception)
        {
            _logger.LogWarning("Redis unavailable — using in-memory cache");
        }
    }

    private bool RedisUp => _redis is { IsConnected: true };

    public async Task<T?> GetAsync<T>(string key)
    {
        if (RedisUp)
        {
            try
            {
                var raw = await _redis!.GetDatabase().StringGetAsync(key);
                return raw.HasValue ? JsonSerializer.Deserialize<T>(raw!) : default;
            }
            catch { /* fall through */ }
        }
        if (_memory.TryGetValue(key, out var hit) && hit.ExpiresAt > DateTime.UtcNow)
            return JsonSerializer.Deserialize<T>(hit.Value);
        _memory.TryRemove(key, out _);
        return default;
    }

    public async Task SetAsync(string key, object value, int ttlSeconds = 60)
    {
        var raw = JsonSerializer.Serialize(value);
        if (RedisUp)
        {
            try
            {
                await _redis!.GetDatabase().StringSetAsync(key, raw, TimeSpan.FromSeconds(ttlSeconds));
                return;
            }
            catch { /* fall through */ }
        }
        _memory[key] = (raw, DateTime.UtcNow.AddSeconds(ttlSeconds));
    }

    public async Task InvalidatePrefixAsync(string prefix)
    {
        if (RedisUp)
        {
            try
            {
                var endpoint = _redis!.GetEndPoints().First();
                var server = _redis.GetServer(endpoint);
                var db = _redis.GetDatabase();
                await foreach (var key in server.KeysAsync(pattern: $"{prefix}*"))
                    await db.KeyDeleteAsync(key);
            }
            catch { /* fall through */ }
        }
        foreach (var key in _memory.Keys.Where(k => k.StartsWith(prefix)))
            _memory.TryRemove(key, out _);
    }
}

File: node-stack/backend/src/cache/cache.service.ts
import { Injectable, Logger, OnModuleDestroy } from '@nestjs/common';
import Redis from 'ioredis';

/**
 * Redis-backed cache with a transparent in-memory fallback so the app
 * remains fully functional when Redis is unavailable (e.g. local dev).
 */
@Injectable()
export class CacheService implements OnModuleDestroy {
  private readonly logger = new Logger(CacheService.name);
  private redis: Redis | null = null;
  private redisHealthy = false;
  private memory = new Map<string, { value: string; expiresAt: number }>();

  constructor() {
    const url = process.env.REDIS_URL || 'redis://localhost:6379';
    try {
      this.redis = new Redis(url, {
        lazyConnect: false,
        maxRetriesPerRequest: 1,
        retryStrategy: (times) => (times > 3 ? null : Math.min(times * 500, 2000)),
      });
      this.redis.on('ready', () => {
        this.redisHealthy = true;
        this.logger.log('Redis connected — caching enabled');
      });
      this.redis.on('error', () => {
        if (this.redisHealthy) this.logger.warn('Redis connection lost — using in-memory cache');
        this.redisHealthy = false;
      });
    } catch {
      this.logger.warn('Redis unavailable — using in-memory cache');
    }
  }

  async get<T>(key: string): Promise<T | null> {
    if (this.redisHealthy && this.redis) {
      try {
        const raw = await this.redis.get(key);
        return raw ? (JSON.parse(raw) as T) : null;
      } catch { /* fall through */ }
    }
    const hit = this.memory.get(key);
    if (hit && hit.expiresAt > Date.now()) return JSON.parse(hit.value) as T;
    this.memory.delete(key);
    return null;
  }

  async set(key: string, value: unknown, ttlSeconds = 60): Promise<void> {
    const raw = JSON.stringify(value);
    if (this.redisHealthy && this.redis) {
      try {
        await this.redis.set(key, raw, 'EX', ttlSeconds);
        return;
      } catch { /* fall through */ }
    }
    this.memory.set(key, { value: raw, expiresAt: Date.now() + ttlSeconds * 1000 });
  }

  async invalidatePrefix(prefix: string): Promise<void> {
    if (this.redisHealthy && this.redis) {
      try {
        const keys = await this.redis.keys(`${prefix}*`);
        if (keys.length) await this.redis.del(...keys);
      } catch { /* fall through */ }
    }
    for (const key of this.memory.keys()) {
      if (key.startsWith(prefix)) this.memory.delete(key);
    }
  }

  onModuleDestroy() {
    this.redis?.disconnect();
  }
}
